### PR TITLE
Adding PyRTL package

### PR DESCRIPTION
**Pull request by mdko**
This PR adds the PyRTL package, a library for hardware design in Python.

I tested it by building and running the added `test_pyrtl.py` file:
Screen Shot 2021-06-05 at 4 15 44 AM

I also ran a local Pyodide console to verify I can import and use it properly:
Screen Shot 2021-06-05 at 4 18 12 AM

### CODE REVIEW

File: packages/pyrtl/test_pyrtl.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            x.next |= a
+            y.next |= b
+        with x > y:
+            x.next |= x - y
+        with y > x:
+            y.next |= y - x
+        with pyrtl.otherwise:
+            done |= True
+
+    gcd <<= x
```
Comment (rth) [inline packages/pyrtl/test_pyrtl.py:25]: That would be a syntax error unless I'm missing something?

File: packages/pyrtl/patches/ctypes_and_renderer.patch [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ from .core import working_block
+ from .wire import Input, Output, Const, WireVector, Register
+@@ -24,6 +22,11 @@ class DllMemInspector(collections.Mapping):
+     """ Dictionary-like access to a hashmap in a CompiledSimulation. """
+ 
+     def __init__(self, sim, mem):
++        try:
++            import ctypes
++        except ImportError:
++            raise PyrtlError('need ctypes installed')
```
Comment (rth) [inline packages/pyrtl/patches/ctypes_and_renderer.patch:33]: Do you really need to catch the error? Failing with the ImportError might be enough?

**Review by rth — COMMENTED** [review]
Thanks @mdko! LGTM aside from the comments below, assuming the test passes.

Do you think it would it would be possible to add this patch upstream, at least the part that moves the ctypes imports inside the functions?

File: packages/pyrtl/patches/ctypes_and_renderer.patch (hunk 2 above)
Comment (mdko) [reply packages/pyrtl/patches/ctypes_and_renderer.patch]: Yes, I think it makes sense to fail with the ImportError in most of the cases except when trying to initialize the CompiledSimulation object; I think it's helpful to have at least one message giving the user an idea what do (i.e. use a different type of simulation).

I've updated the patch to just `import ctypes` where needed like you suggest, and now only display the additional PyrtlError in the object initialization.

File: packages/pyrtl/test_pyrtl.py (hunk 1 above)
Comment (mdko) [reply packages/pyrtl/test_pyrtl.py:26]: This isn't a syntax error. At that point, `done` and `gcd` are WireVectors, which have `__ilshift__` and `__ior__` (. (among many other methods) defined to do the behind-the-scenes magic of creating netlists behind the scenes

**Comment by mdko** [comment]
First, thank you for your quick feedback! I've updated the patch (see comments above) and ran black on the test.

If you're okay with it, I'd prefer to leave the upstream code as is with respect to the `ctypes` import, since `ctypes` is part of the Python standard library and can normally be expected to be available. I feel like patching is the appropriate way to fit this into the framework.

**Review by rth — APPROVED** [review]
Thanks @mdko LGTM.

**Pull request merged**